File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/grouplist/GroupListPresenter.java
package com.mifos.mifosxdroid.online.grouplist;

import com.mifos.api.DataManager;
import com.mifos.mifosxdroid.base.BasePresenter;
import com.mifos.objects.group.CenterWithAssociations;
import com.mifos.objects.group.GroupWithAssociations;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;


/**
 * Created by Rajan Maurya on 06/06/16.
 */
public class GroupListPresenter extends BasePresenter<GroupListMvpView> {

    private final DataManager mDataManager;
    private CompositeDisposable compositeDisposable;

    @Inject
    public GroupListPresenter(DataManager dataManager) {
        mDataManager = dataManager;
    }


    @Override
    public void attachView(GroupListMvpView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        if (compositeDisposable != null) compositeDisposable.clear();
    }

    public void loadGroupByCenter(int id) {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        if (compositeDisposable != null) compositeDisposable.clear();
        compositeDisposable = mDataManager.getGroupsByCenter(id)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<CenterWithAssociations>() {
                    @Override
                    public void onComplete() {
                        getMvpView().showProgressbar(false);
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showFetchingError("Failed to load GroupList");
                    }

                    @Override
                    public void onNext(CenterWithAssociations centerWithAssociations) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showGroupList(centerWithAssociations);
                    }
                });
    }

    public void loadGroups(int groupid) {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        if (compositeDisposable != null) compositeDisposable.clear();
        compositeDisposable = mDataManager.getGroups(groupid)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<GroupWithAssociations>() {
                    @Override
                    public void onComplete() {
                        getMvpView().showProgressbar(false);
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showFetchingError("Failed to load Groups");
                    }

                    @Override
                    public void onNext(GroupWithAssociations groupWithAssociations) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showGroups(groupWithAssociations);
                    }
                });
    }


}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/grouplist/GroupListPresenter.kt
package com.mifos.mifosxdroid.online.grouplist

import com.mifos.api.DataManager
import com.mifos.mifosxdroid.base.BasePresenter
import com.mifos.objects.group.CenterWithAssociations
import com.mifos.objects.group.GroupWithAssociations
import rx.Subscriber
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by Rajan Maurya on 06/06/16.
 */
class GroupListPresenter @Inject constructor(private val mDataManager: DataManager) : BasePresenter<GroupListMvpView?>() {
    private var mSubscription: Subscription? = null
    override fun attachView(mvpView: GroupListMvpView?) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        if (mSubscription != null) mSubscription!!.unsubscribe()
    }

    fun loadGroupByCenter(id: Int) {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        if (mSubscription != null) mSubscription!!.unsubscribe()
        mSubscription = mDataManager.getGroupsByCenter(id)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<CenterWithAssociations?>() {
                    override fun onCompleted() {
                        mvpView!!.showProgressbar(false)
                    }

                    override fun onError(e: Throwable) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showFetchingError("Failed to load GroupList")
                    }

                    override fun onNext(centerWithAssociations: CenterWithAssociations?) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showGroupList(centerWithAssociations)
                    }
                })
    }

    fun loadGroups(groupid: Int) {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        if (mSubscription != null) mSubscription!!.unsubscribe()
        mSubscription = mDataManager.getGroups(groupid)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<GroupWithAssociations?>() {
                    override fun onCompleted() {
                        mvpView!!.showProgressbar(false)
                    }

                    override fun onError(e: Throwable) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showFetchingError("Failed to load Groups")
                    }

                    override fun onNext(groupWithAssociations: GroupWithAssociations?) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showGroups(groupWithAssociations)
                    }
                })
    }

}